src/Mvc/Mvc.Components.Prerendering/test/HtmlHelperComponentPrerenderingExtensionsTests.cs

```diff
@@ -0,0 +1,470 @@
+                {
+                    await Task.Yield();
+                    throw new InvalidOperationException("Threw an exception asynchronously");
+                }
+            }
+        }
+
+        private class GreetingComponent : ComponentBase
+        {
+            [Parameter] public string Name { get; set; }
+
+            protected override void OnParametersSet()
+            {
+                base.OnParametersSet();
+            }
+
+            protected override void BuildRenderTree(RenderTreeBuilder builder)
+            {
+                var s = 0;
+                base.BuildRenderTree(builder);
+                builder.OpenElement(s++, "p");
+                builder.AddContent(s++, $"Hello {Name}!");
+                builder.CloseElement();
+            }
+        }
+
+        private class AsyncComponent : ComponentBase
+        {
+            private static WeatherRow[] _weatherData = new[]
+            {
+                new WeatherRow
+                {
+                    DateFormatted = "06/05/2018",
+                    TemperatureC = 1,
+                    Summary = "Freezing",
+                    TemperatureF = 33
+                },
+                new WeatherRow
+                {
+                    DateFormatted = "07/05/2018",
+                    TemperatureC = 14,
+                    Summary = "Bracing",
+                    TemperatureF = 57
+                },
+                new WeatherRow
+                {
+                    DateFormatted = "08/05/2018",
+                    TemperatureC = -13,
+                    Summary = "Freezing",
+                    TemperatureF = 9
+                },
+                new WeatherRow
+                {
+                    DateFormatted = "09/05/2018",
+                    TemperatureC = -16,
+                    Summary = "Balmy",
+                    TemperatureF = 4
+                },
+                new WeatherRow
+                {
+                    DateFormatted = "10/05/2018",
+                    TemperatureC = 2,
+                    Summary = "Chilly",
+                    TemperatureF = 29
+                }
+            };
+
+            public class WeatherRow
+            {
+                public string DateFormatted { get; set; }
+                public int TemperatureC { get; set; }
+                public string Summary { get; set; }
+                public int TemperatureF { get; set; }
+            }
+
+            public WeatherRow[] RowsToDisplay { get; set; }
+
+            protected override async Task OnParametersSetAsync()
+            {
+                // Simulate an async workflow.
+                await Task.Yield();
+                RowsToDisplay = _weatherData;
+            }
+
+            protected override void BuildRenderTree(RenderTreeBuilder builder)
+            {
+                base.BuildRenderTree(builder);
+                var s = 0;
+                builder.OpenElement(s++, "table");
```

Comment:
Please don't do this `s++` thing! Community members often find it difficult to understand what's problematic about it ([description](https://gist.github.com/SteveSandersonMS/ec232992c2446ab9a0059dd0fbc5d0c3)), so it will be extra-confusing if we do it ourselves.

Also could we make this `GreetingComponent` produce far simpler output so there's not anywhere near as much code here?

Reply:
done